packages/react-ui-core/src/AutoSuggestField/AutoSuggestField.js

```diff
@@ -0,0 +1,216 @@
+    submitOnSelection: PropTypes.bool,
+    visible: PropTypes.bool,
+    suggestions: PropTypes.node,
+    value: PropTypes.string,
+    valueSelector: PropTypes.func,
+  }
+
+  static defaultProps = {
+    theme: {},
+    anchorField: {},
+    value: '',
+    onSubmit: () => { },
+    onInput: () => { },
+    onSelection: () => { },
+    valueSelector: value => value,
+    submitOnSelection: true,
+    visible: false,
+  }
+
+  constructor(props) {
+    super(props)
+    const { value, visible, onInput } = props
+
+    this.onInput = debounce(onInput, 300)
+    this.state = {
+      value,
+      visible,
+    }
+  }
+
+  componentDidMount() {
+    document.addEventListener('keydown', this.onKeyDown)
+  }
+
+  componentWillUnmount() {
+    document.removeEventListener('keydown', this.onKeyDown)
+  }
+
+  @autobind
+  onKeyDown() {
+    const { suggestions } = this.props
+    const { visible } = this.state
+
+    switch (event.keyCode) {
+      case ESCAPE:
+        this.handleClear()
+        break
+      case ENTER:
+        if (!visible || !suggestions || !suggestions.length) {
+          this.handleSubmit(this.state.value)
+        }
+        break
+      default:
+    }
+  }
+
+  updateValueAndClose(value, cb = () => { }) {
+    this.setState({
+      value,
+      visible: false,
+    }, cb)
+  }
+
+  @autobind
+  handleInput(event) {
+    const value = event.target.value
+    this.setState({ value, visible: true })
+    this.onInput(value)
+  }
+
+  @autobind
+  handleSubmit() {
+    this.props.onSubmit(this.state.value)
+    this.updateValueAndClose(this.state.value)
+  }
+
+  @autobind
+  handleSuggestionSelection(value) {
+    const {
+      onSubmit,
+      onSelection,
+      submitOnSelection,
+      onInput,
+      valueSelector,
+    } = this.props
+
+    onInput(value)
+    onSelection(value)
+
+    this.updateValueAndClose(
+      valueSelector(value),
+      () => {
+        if (submitOnSelection) onSubmit(this.state.value)
+      }
+    )
+  }
+
+  @autobind
+  handleVisibilityChange(visible) {
+    this.setState({
+      visible,
+    })
+  }
+
+  @autobind
+  handleClear() {
+    const { onAfterClear } = this.props
+    this.updateValueAndClose('')
+    if (onAfterClear) onAfterClear()
+  }
+
+  renderButton(button, type, props) {
+    if (!button) return null
+
+    return createElement(...parseArgs(button, Button, {
+      ...props,
+      onClick: this[`handle${capitalize(type)}`],
+      'data-tid': `autosuggest-${type}-button`,
+    }))
+  }
+
+  renderAnchorField() {
+    return {
+      ...this.props.anchorField,
+      value: this.state.value,
+      anchor: Field,
+      onChange: this.handleInput,
+    }
+  }
+
+  render() {
+    const {
+      theme,
+      clearButton,
+      submitButton,
+      anchorField,
+      suggestions,
+      className,
+      onSubmit, // eslint-disable-line no-unused-vars
+      onItemMouseOver,
+      onSelection, // eslint-disable-line no-unused-vars
+      valueSelector,
```

Comment:
Why is onSubmit and onSelection not used?

Reply:
They aren't used here, but they are used in `handleSelectionSuggestion` and `handleSubmit` so they shouldn't be passed on to the `Dropdown`.

Reply:
We need to make sure they don't get passed down through spread props, so we're pulling them out even though we aren't using them in this scope.

Reply:
Ah, that makes sense. You are pulling them out so they don't get passed down.